src/toast/ops/totalconvolve.py

```diff
@@ -0,0 +1,654 @@
+
+        if verbose:
+            timer.report_clear(f"convolve detector {det}")
+            timer.report_clear(f"extract convolved data for {det}")
+
+        del convolver
+        convolved_data *= 0.5  # FIXME: not sure where this factor comes from
+        return convolved_data
+
+    def calibrate_signal(self, data, det, beam, convolved_data, verbose):
+        """By default, libConviqt results returns a signal that conforms to
+        TOD = (1 + epsilon) / 2 * intensity + (1 - epsilon) / 2 * polarization.
+
+        When calibrate = True, we rescale the TOD to
+        TOD = intensity + (1 - epsilon) / (1 + epsilon) * polarization
+        """
+        if not self.calibrate: # or beam.normalized():
+            return
+          
+        timer = Timer()
+        timer.start()
+        offset = 0
+        for obs in data.obs:
+            if det not in obs.local_detectors:
+                continue
+            focalplane = obs.telescope.focalplane
+            epsilon = self._get_epsilon(focalplane, det)
+            # Make sure detector data output exists
+            obs.detdata.ensure(self.det_data, detectors=[det])
+            # Loop over views
+            views = obs.view[self.view]
+            for view in views.detdata[self.det_data]:
+                nsample = len(view[det])
+                convolved_data[offset : offset + nsample] *= 2 / (1 + epsilon)
+                offset += nsample
+        if verbose:
+            timer.report_clear(f"calibrate detector {det}")
+        return
+
+    def save(self, data, det, convolved_data, verbose):
+        """Store the convolved data."""
+        timer = Timer()
+        timer.start()
+        offset = 0
+        for obs in data.obs:
+            if det not in obs.local_detectors:
+                continue
+            # Loop over views
+            views = obs.view[self.view]
+            for view in views.detdata[self.det_data]:
+                nsample = len(view[det])
+                view[det] += convolved_data[offset : offset + nsample]
+                offset += nsample
+        if verbose:
+            timer.report_clear(f"save detector {det}")
+        return
+
+    def _finalize(self, data, **kwargs):
+        return
+
+    def _requires(self):
+        req = self.detector_pointing.requires()
+        req["meta"].extend([self.noise_model, self.pixel_dist, self.covariance])
+        req["shared"] = [self.boresight]
+        if self.shared_flags is not None:
+            req["shared"].append(self.shared_flags)
+        if self.det_flags is not None:
+            req["detdata"].append(self.det_flags)
+        if self.view is not None:
+            req["intervals"].append(self.view)
+        return req
+
+    def _provides(self):
+        prov = self.detector_pointing.provides()
+        prob["detdata"].append(self.det_data)
+        return prov
+
+    def _accelerators(self):
+        return list()
+
+
+class SimWeightedTotalconvolve(SimTotalconvolve):
```

Comment:
I thought one of the advantages of the totalconvolve code was that it could natively support the HWP without the kind of trickery we have resort to. Could these two operators be merged into one?

Reply:
Yes, I think so. At the moment my goal is just to produce a "drop-in" (as far as possible) replacement for libconviqt, which should make comparison and validation easier.

Reply:
Done.